### DISCUSSION TITLE
Guidance on downloading whole websites

### DISCUSSION THREAD

**Discussion by BradKML** [rajatomar788/pywebcopy#112] [Q&A]
Intended goal: Downloading whole "self-help" websites to hot-plugging individual page entries into LLMs, and use it to cluster webpage content.

Requirement:
1. Be able to scrape websites quickly (since most of the sites uses little to no JS, `open_in_browser` would not make much sense, similar to how `bs4` is always faster than `selenium` for other scraper solutions)
2. Scrape them in a stealthy manner (`bypass_robots` may be an issue, but `delay=None` would be good
3. Having ability to scale and centralize (`threaded` between multiple IPs/computers would be better than just concurrently downloading from multiple websites)
4. Be able to read scraped pages as individual units for HTML extractors (e.g. [Trafilatura](https://github.com/scrapinghub/article-extraction-benchmark)) and language libraries (e.g. [SBERT](https://www.sbert.net/) or [SpaCy-USE](https://github.com/MartinoMensio/spacy-universal-sentence-encoder))

Questions:
1. Is it possible to "smell" if the server might block scrapers? Also, what are the ways to work around it (e.g. user agents)? 
2. How are "no-cobbler" like features used in this context for suspended and resumed downloads? https://www.linuxjournal.com/content/downloading-entire-web-site-wget
3. How can one prevent whole-site scraping to leech into other unrelated websites? https://gist.github.com/crittermike/fe02c59fed1aeebd0a9697cf7e9f5c0c
4. How can individual files (either file path OR list-of-strings) be listed as output for piping into other tools?
5. How can this process be broken down into concurrent parts (between websites/servers), asynchronous procedures (thread-like within the same website), or pauseable parts?
6. What would happen if you want to capture a blog with daily updated posts, and have this operation run every month? (likely the multimedia URLs and files would not change much, not worth duplicating the work)

**Comment by rajatomar788** [comment]
> Is it possible to "smell" if the server might block scrapers? Also, what are the ways to work around it (e.g. user agents)?

No there is no way anyone can smell blocking just like in the messenger apps. 

But you can definitely change the user-agent string

```

from pywebcopy.config import get_config
config = get_config(url, project_folder, project_name, bypass_robots, debug, delay, threaded)
webpage = config.create_page()

# here you have the ready webpage instance to do anything
webpage.session.header = anything
webpage.get(url)
    
webpage.save_complete(pop=open_in_browser)

```

**Reply by BradKML** [reply]
What about checking the robot file and see if they have a "no scrape" policy (probably the site have other cloudFlare-esque roadblocks too)? Would that work?

**Reply by rajatomar788** [reply]
It does check the robot file, that is what `bypass_robots` option is for.

**Reply by BradKML** [reply]
More in the sense of adjusting crawling speed/tactics based on the robot file rather than merely quitting. As a side option, play by the robot rules, but have multiple IPs distributing the load to increase efficiency (multiple thread on a single IP seems like a different issue).
How hard does `bypass_robots` go, do you have to adjust pause time manually?

**Reply by rajatomar788** [reply]
Multiple IPs and all that is the job of the user not the library. How can library generate public IP addresses without having complete control of user device and network. Would you like that?

**Reply by BradKML** [reply]
Assuming that the user has access to multiple IPs, how could the software adapt to it?

**Comment by rajatomar788** [comment]
> How are "no-cobbler" like features used in this context for suspended and resumed downloads? https://www.linuxjournal.com/content/downloading-entire-web-site-wget

Yes there is this equivalent config key `overwrite` which you could set to False to prevent redownloading of files.

**Reply by BradKML** [reply]
Thanks for the note, in this case there was an issue by someone to do conditional check for overwrites through checking headers.
https://github.com/rajatomar788/pywebcopy/issues/41
Some other people proposed either checksum (very fragile to ad and dynamic websites) or article metadata extraction (see [example](https://trafilatura.readthedocs.io/en/latest/corefunctions.html#extract-metadata), but platform-dependent).

**Reply by rajatomar788** [reply]
As you said it is very case specific and bandwidth heavy so not worth trouble in one off downloads.

**Reply by BradKML** [reply]
When it is dealing with whole blog archiving (plus updating), I do think it could be of use.

**Reply by rajatomar788** [reply]
You can write a mod pull request but right now it does its job light and clean.

**Comment by rajatomar788** [comment]
> How can one prevent whole-site scraping to leech into other unrelated websites? https://gist.github.com/crittermike/fe02c59fed1aeebd0a9697cf7e9f5c0c

Well you have to manually blacklist or whitelist domains in the session object as of pywebcopy 7.0.2, this is a design limitation.

**Reply by BradKML** [reply]
Would kindly request an example of using it since the documentation is a bit sparse. Trying to hop from WGet to PyWebCopy take a bit of a learning curve.

**Reply by rajatomar788** [reply]
```
from pywebcopy.config import get_config
config = get_config(url, project_folder, project_name, bypass_robots, debug, delay, threaded)
webpage = config.create_page()
 

webpage.session.domain_blacklist = your_list[]

webpage.get(url)
webpage.save_complete(pop=open_in_browser)
```

**Comment by rajatomar788** [comment]
> How can individual files (either file path OR list-of-strings) be listed as output for piping into other tools?

The record of all the urls along with their downloaded path is stored in a `pywebcopy.schedulers.RecentOrderedDict` in `pywebcopy.schedulers.Scheduler` objects `index` attribute.

```
from pywebcopy.config import get_config
config = get_config(url, project_folder, project_name, bypass_robots, debug, delay, threaded)
webpage = config.create_page()

# here you have the ready webpage scheduler to hook up you custom version of a RecentOrderedDict 

webpage.scheduler.index = your_callback_wrapped_RecentOrderedDict


webpage.get(url)
webpage.save_complete(pop=open_in_browser)
```

**Reply by BradKML** [reply]
I guess the `RecentOrderedDict` will be URL-filepath pairs that can be iterated like an `OrderedDict` Python object, got it.

The flow would be create config, then make webpage crawler object, get the data, and `save_complete`? If the crawling is not completed in the background, what would `save_complete` do in the meantime? Would they provide an alert when the process is done?

Just to confirm, this is saving one page, or a whole site crawl (and getting list of HTML files)? It is kind of weird to "open in browser" if you saved a website with 100+ pages. I guess `pop=False` then?

**Reply by rajatomar788** [reply]
@BrandonKMLee save_complete just starts the process so you would need to call it. In case of pop you can do whatever gives better results.

**Reply by BradKML** [reply]
Does PyWebCopy run on bs4-like (see also scrapy) browserless scraping or Selenium-like browsered scraping? Assuming the websites does not have enough JS to block off article downloads, would `pop=False` be more suitable?

**Reply by rajatomar788** [reply]
Pywebcopy uses lxml library for browser less scraping. Pop just opens the first page after completion so i don't know.

**Reply by BradKML** [reply]
If it is LXML it would operate similar to BeautifulSoup. In that case are there code samples for doing the same thing but for whole websites? Following [this documentation](https://github.com/rajatomar788/pywebcopy/wiki/How-To#how-to---clone-whole-websites) only downloaded one page instead of the intended effect.

**Reply by rajatomar788** [reply]
That documentation link is from 2019. Use the latest Readme file documentation.

**Reply by BradKML** [reply]
Got it. Thanks for the note, hope to see a more comprehensive doc soon.

**Comment by rajatomar788** [comment]
> How can this process be broken down into concurrent parts (between websites/servers), asynchronous procedures (thread-like within the same website), or pauseable parts?

You have the `threaded=True` config option.

**Comment by rajatomar788** [comment]
> What would happen if you want to capture a blog with daily updated posts, and have this operation run every month? (likely the multimedia URLs and files would not change much, not worth duplicating the work)

What happens when you visit a social media site daily or even hourly? Nothing.
